src/execute.rs

```diff
@@ -0,0 +1,654 @@
+        let amount_1_delta = calc_amount_one_delta(
+            Decimal256::from_str(pool.current_tick_liquidity.as_str()).unwrap(),
+            pool_cur_sqrt_price,
+            sqrt_price_next,
+            false,
+        );
+
+        let amt0_to_swap = token_provided_dec
+            .checked_mul(asset_1_ratio_new.checked_add(spread_factor)?)?
+            .to_uint_floor();
+        let amt0_for_pos = token_initially_provided
+            .amount
+            .checked_sub(amt0_to_swap.try_into()?);
+        let sqrt_price_next = get_next_sqrt_price_from_amount0_in_round_up(
+            Decimal256::from_str(pool.current_tick_liquidity.as_str()).unwrap(),
+            pool_cur_sqrt_price,
+            Decimal256::from_str(amt0_to_swap.to_string().as_str()).unwrap(),
+        );
+        let price_next = sqrt_price_to_price(sqrt_price_next);
+
+        // Determine the tick this will bring us to
+        let tick_next = price_to_tick(deps.storage, price_next)?;
+        deps.api.debug(&format!("tick_next {tick_next:?}"));
+        deps.api.debug(&format!(
+            "final position pair: amt0 {amt0_for_pos:?}, amt1 {amount_1_delta:?}"
+        ));
+        return Ok(Coin {
+            denom: token_to_swap.denom,
+            amount: amt0_to_swap.try_into()?,
+        });
+    } else if token_1_in {
+        let amount_0_delta = calc_amount_zero_delta(
+            Decimal256::from_str(pool.current_tick_liquidity.as_str()).unwrap(),
+            pool_cur_sqrt_price,
+            sqrt_price_next,
+            true,
+        );
+        deps.api
+            .debug(&format!("amount_0_delta {amount_0_delta:?}"));
+        let amt1_to_swap = token_provided_dec
+            .checked_mul(asset_0_ratio_new.checked_add(spread_factor)?)?
+            .to_uint_floor();
+        let amt1_for_pos = token_initially_provided
+            .amount
+            .checked_sub(amt1_to_swap.try_into()?);
+        let sqrt_price_next = get_next_sqrt_price_from_amount1_in_round_down(
+            Decimal256::from_str(pool.current_tick_liquidity.as_str()).unwrap(),
+            pool_cur_sqrt_price,
+            Decimal256::from_str(amt1_to_swap.to_string().as_str()).unwrap(),
+        );
+        let price_next = sqrt_price_to_price(sqrt_price_next);
+
+        // Determine the tick this will bring us to
+        let tick_next = price_to_tick(deps.storage, price_next)?;
+        deps.api.debug(&format!("tick_next {tick_next:?}"));
+        deps.api.debug(&format!(
+            "final position pair: amt0 {amount_0_delta:?}, amt1 {amt1_for_pos:?}"
+        ));
+        return Ok(Coin {
+            denom: token_to_swap.denom,
+            amount: amt1_to_swap.try_into()?,
+        });
+    } else {
+        return Err(ContractError::DenomNotInPool {
+            provided_denom: token_to_swap.denom,
+        });
+    }
+}
+
+fn sqrt_price_to_price(sqrt_price: Decimal256) -> Decimal256 {
+    return sqrt_price * sqrt_price;
+}
+
+// fn iterative_naive_approach_new(
```

Comment:
Note to reviewer: This was an initial attempt at iterating to get the perfect ratio, but after discussing with Dev I scrapped this for now